ostd/src/sync/rcu/monitor.rs

```diff
@@ -27,120 +26,54 @@ impl RcuMonitor {
     /// The singleton instance is globally accessible via the `RCU_MONITOR`.
     pub fn new() -> Self {
         Self {
-            is_monitoring: AtomicBool::new(false),
-            state: SpinLock::new(State::new()),
+            cur_queue_dropping: AtomicBool::new(false),
+            cpus_passed_quiescent: AtomicCpuSet::new(CpuSet::new_empty()),
+            cur_waitqueue: SpinLock::new(VecDeque::new()),
+            next_waitqueue: SpinLock::new(VecDeque::new()),
         }
     }
 
     pub(super) unsafe fn pass_quiescent_state(&self) {
-        // Fast path
-        if !self.is_monitoring.load(Relaxed) {
-            return;
-        }
-
-        // Check if the current GP is complete after passing the quiescent state
-        // on the current CPU. If GP is complete, take the callbacks of the current
-        // GP.
-        let callbacks = {
-            let mut state = self.state.disable_irq().lock();
-            if state.current_gp.is_complete() {
-                return;
-            }
-
-            state.current_gp.pass_quiescent_state();
-            if !state.current_gp.is_complete() {
+        let preempt_guard = disable_preempt();
+        let current_cpu = preempt_guard.current_cpu();
+
+        self.cpus_passed_quiescent
+            .add(current_cpu, Ordering::Relaxed);
+
+        if self.cpus_passed_quiescent.load().is_full() {
+            if self
+                .cur_queue_dropping
+                .compare_exchange(false, true, Ordering::AcqRel, Ordering::Acquire)
```

Comment:
Would you mind removing the commit "Make RCU locks less contended" from this PR? This commit is not sound.

Example:
```
---- CPU 1 ----                                  ---- CPU 2 ---

cpus_passed_quiescent == FULL

cur_queue_dropping: false -> true

[ dropping... ]
                                                       
                                                 cpus_passed_quiescent == FULL

cpus_passed_quiescent: FULL -> EMPTY

cur_queue_dropping: true -> false

                                                 cur_queue_dropping: false -> true

						 [ dropping... ]
						   ^
						   |
						  This will drop the items in the _current_ GP,
						  which is not sound.
```

Before you open this PR, I have been thinking for a while about how to properly optimize the `pass_quiescent_state` method to make it lockless and even atomic-operation-less in most common cases. I have some ideas, but haven't finalized them in any doc or code yet.

So if this commit isn't that important right now and if you don't mind removing this commit from this PR, maybe I can try my ideas after this PR is merged.

Reply:
I am also concerned with the safety and efficiency of this implementation of `pass_quiescent_state`.

Reply:
Yes. Thanks for pointing it out.